---
layout: post
title: Basic Concepts
description: >
   temlplate, const, exception handling, inline function, 포인터와 참조자
sitemap: false
hide_last_modified: true
---

<br>



> 모르면 안될 필수적이고 기본적인 개념들 정리



---

<br>



# template



<br>



>template는 간단하게 생각하면 **함수나 클래스의 data type을 자유롭게 해주는 도구**로, 함수의 리턴형이나 파라미터로 전달받는 변수들의 data tyoe을 미리 정해놓는 것이 아닌, **함수 호출시에 파라미터로 data type을 전달받아 해당 형으로 리턴**해주는 것이다.  <br>
>
>쉽게 말하면 함수나 클래스를 개별적으로 다시 작성하지 않아도 여러 type으로 사용할 수 있도록 만들어주는 '**틀**' 이라고 생각하면 된다.



<br>



+ 사용법



<br>



ex) 

```cpp

int sum(int a, int b){

    return a + b;

}

double sum(double a, double b){

    return a + b;

}

```



<br>



+ 예를 들어, 위와 같이 두개의 data type으로 동일한 계산을 수행하고 싶을 때, 아래와 같이 함수를 template로 만들어 놓으면 함수를 여러번 작성할 필요 없이 편하게 사용할 수 있다.



<br>



```cpp

template <typename T>

T sum(T a, T b){

    return a + b;

}

```



<br>



+ 조금 더 완성된 코드를 살펴보자.



<br>



### 예시 1)



```cpp

#include<iostream>

#include<string>

 

using namespace std;

 

template <typename T>

T sum(T a, T b){

    return a + b;

}

 

int main(void) {

    int a=1, b =2;

    

    double d1 = 2.2;

    double d2 = 3.3;

 

    string s1 = "Show me ";

    string s2 = "The Money 6";

    

    cout << "int 합 : " << sum<int>(a, b) << endl;

    cout << "double 합 : " << sum<double>(d1, d2) << endl;

    cout << "string 합 : " << sum<string>(s1, s2) << endl;

    

    return 0;

}



```



<br>





### 예시 2)



```cpp

#include <iostream>

#include <string>

 

using namespace std;

 

template <class T1, class T2>

void printAll(T1 a, T2 b){

    cout << "T1 : " << a << endl;

    cout << "T2 : " << b << endl;

};

 

int main(void){

    string s1 = "Dok2" ;

    string s2 = "On my way." ;

 

    int num1 = 27;

    int num2 = 35;

 

    double d1 = 3.14;

    double d2 = 36.5;

 

    cout << "[string, string]" << endl;

    printAll(s1, s2);                   //printAll<string, string>(s1,s2);

 

    cout << "[string, int]" << endl;

    printAll(s1, num1);                 //printAll<string, int>(s1,num1);

 

    cout << "[int, int]" << endl;

    printAll(num1, num2);

 

    cout << "[int, double]" << endl;

    printAll(num1, d1);

 

    cout << "[double, double]" << endl;

    printAll(d1, d2);

 

    cout << "[double, string]" << endl;

    printAll(d1, s1);

 

    return 0;

}

```



<br>



> template로 사용하는 변수가 2개 이상일 경우에는 함수이름 뒤의 <> 안에 자료형을 명확하게 표기하지않고, 컴파일러가 자동적으로 판단하게 된다.
>



<br>



+ 위의 코드가 이해된다면, template는 제대로 사용할 수 있을 것이다.



<br>



---

<br>



# const



<br>



>const는 말 그대로 '상수'를 뜻하는데, 이는 함수 앞에도, 변수 앞에도 붙을 수 있다. <br>
>
>상수라는 말이 어떠한 의미인가? 변하지 않는다는 뜻 이다.  <br>
>
>한마디로 **특정 함수 앞에 const로 선언이 되어있다면, 해당 함수의 리턴값은 후에 처리하면서 변경할 수 없다는 말이고, 특정 변수 앞에 const가 붙어있다면, 해당 변수의 값은 상수화되어 앞으로 계산 과정에서 변할 수 없다는 뜻**이다. 



<br>



>왜 사용해야되는걸까? <br>
>
>일종의 **안전장치**라고 생각하면 좋다. <br>
>
>우리는 프로그램을 작성하는 도중에 실수로 상수화 시키고 싶었던 변수의 값을 건드려도 알 길이 없다. <br>
>
>하지만 const 키워드를 붙여두면 해당 변수의 값을 변경하려 시도했을 때 에러가 발생하며 해당 변수의 값이 달라짐을 인지할 수 있게 되는 것이다. 쉽게 생각하면 이러하다. 



<br>



>다만 주의할 점은, const 변수는 반드시 '**선언 시 초기화**'를 해주어야 컴파일 에러가 발생하지 않는다. <br>
>
>그래서 class의 멤버 변수를 const로 선언 시에는 반드시 초기화 리스트(Initialize List)를 사용해야 한다.



<br>



---

<br>





# Exception Handling (예외 처리) : try catch / throw



<br>



## try catch



<br>



>예외처리를 if문으로 해도 되는데 왜 굳이 try catch 를 쓰는 것일까? <br>
>
>if문으로 처리해도 상관은 없다. 다만 해당 코드가 **예외처리를 위한 파트임을 보다 명시적으로 나타내고자** try catch문을 사용한다고 생각하면 편하겠다.



<br>



+ 기본 문법



<br>



```cpp

try { 

    // 예외 발생 예상 지역

    if (예외 조건) throw 예외 객체; // exception 발생 -> throw

}

catch (예외 객체 1) { // throw된 exception을 catch

    // 예외 처리1

}

catch (예외 객체 2) {

    // 예외 처리 2

}

catch (...) {

    // 

}

```



<br>



+ try 블록에서 예외가 발생하면, 예외가 발생한 지점 이후를 실행하는 것이 아니라, catch 블록 이후가 실행된다.



+ 예시로는 아래와 같이 처리할 수 있다.



<br>



```cpp

#include <iostream>

using namespace std;



int main() {

    try {

        int dividend, divisor;

        

        cout << "Enter dividend: ";

        cin >> dividend;

        

        cout << "Enter divisor: ";

        cin >> divisor;

        

        if (divisor == 0) {

        // std::runtime_error 예외 던지면서 메시지 전달

            throw runtime_error("Division by zero is not allowed.");

        }

        

        double result = static_cast<double>(dividend) / divisor;

        cout << "Result: " << result << endl;

    } catch (const std::exception& e) {

    // 예외 객체를 통해 메시지 출력

        cout << "An exception occurred: " << e.what() << endl;

    }

  

    return 0;

}

```



<br>



> 이 외에도 C++ 표준 라이브러이에서 std::exception 을 상속받는 예외 클래스가 많으니 아래를 살펴보자.



<br>



+ **std::runtime_error** : 프로그램 실행 중 발생하는 오류

+ **std::logic_erro**r : 논리 오류. 예를 들어, 잘못된 인자를 전달한 경우 등

+ **std::invalid_argument** : 잘못된 인수를 전달한 경우 발생하는 예외 클래스

+ **std::out_of_range** : 범위를 벗어난 인덱스나 값에 접근한 경우 발생하는 예외 클래스

+ **std::bad_alloc** : 동적 메모리 할당에 실패한 경우 발생하는 예외 클래스



<br>



> excepton 클래스 관련 내용 출처 : https://eteo.tistory.com/584



<br>



---

<br>



# inline 함수



<br>



>일반 함수는 프로그램 내에서 함수를 호출하게 되면, 호출 시점의 주소값(명령어의 주소값)을 스택에 저장해놓고 함수의 주소로 이동했다가 내부적으로 처리를 완료하면 다시 호출 시점의 주소로 복귀하여 계산된 값을 리턴하는 형식으로 실행된다.  <br>
>
>이에서 유추할 수 있듯이 함수를 충분히 많은 횟수만큼 호출하게 되면 프로그램의 실행속도가 상당히 저하됨을 느낄 수 있을 것이다. 

 

<br>



>이를 보완하기 위한 것이 inline 함수인데, inline 함수는 일반 함수와 달리 **함수 내부의 코드를 함수를 호출한 시점에 그대로 대입시켜 넣는다.** <br>
>
>때문에 실행 위치가 왔다갔다 하며 시간을 허비할 필요 없이 다수의 **함수 호출을 보다 빠르게 처리**할 수 있게 되는 것이다.

 

<br>



>물론 장점만 존재하지는 않는다.  <br>
>
>앞서 설명한 내용을 보면 인라인 함수의 사본을 반복해서 붙여넣는 작업을 하기 때문에 **메모리 사용 측면에서 일반 함수에 비해 비효율적**으로 변할 수도 있다. 



<br>

 

>때문에 주로 여러번 호출되지만 함수의 내용이 보다 간결한 함수들을 inline 함수로 만들어 사용하는 것이 바람직 하다고 생각된다.



<br>

  

---

<br>



# 포인터와 참조자(참조변수)



<br>



### 포인터와 참조자를 아직도 헷갈려하면 안된다. 큰일난다.



<br>



+ 포인터 : 객체의 '**주소**'를 가리키는 방식

+ 참조자 : 객체의 '**값**'을 참조하는 방식



<br>



> point) 참조자는 '별명'이라고 생각하라는 설명을 봤었는데, 정말 적합한 설명이라 느낀다. 참조자는 어떤 변수의 '별명'이다.



<br>



>어떨 때 사용해 ? <br>
>
>포인터는 특정 변수의 값에 진짜 접근하기 위해 (값만 가져오는 것이 아니라) 사용되고, 참조자는 주로 함수에 특정 변수를 파라미터로 전달할 때 사용된다.  <br>
>
>포인터는, 참조자도 한 값이 변하면 다른 한 값도 변한다. 

 

<br>



>그러면 포인터랑 참조자랑 뭐가 다르냐 싶겠지만, 참조자는 한번 참조하면 대상 변경이 불가능하여 포인터보다 확실하게 안전하다.  <br>
>
>하지만 반드시 포인터를 사용해야 하는 경우도 있는데, 참조자는 선언과 동시에 초기화되어야 하기 때문에 포인터와 달리 NULL 값이 허용되지 않는다.  <br>
>
>따라서 함수의 파라미터로 NULL값 을 전달하거나 NULL값을 리턴받는 경우에는 포인터를 사용해야한다. 



<br>



+ 실제 코드로 살펴보자. 



<br>



## 1. Call By Value



<br>



```cpp

void CallByValue (int value) {

    value = 20;

    cout << value << endl;

    cout << &value << endl;

}



int main() {

    int num = 10;

    cout << num << endl;

    cout << &num << endl;

    return 0;

}

```



<br>



+ 위의 코드를 실행시키면 어떤 결과가 나올까?



<br>



>20 <br>
>
>000000593B51FAE0 <br>
>
>10 <br>
>
>000000593B51FB04



<br>



>이처럼 '값'에 의한 전달을 하기 때문에, num에 저장된 10이라는 '값'을 복사하여 CallByValue()에 전달하고, 복사되어 전달받은 값을 value라는 변수에 저장하는 것이다.  <br>
>
>때문에 함수가 종료되면 value라는 변수가 소멸됨에 따라 복사+전달 받은 값도 소멸된다.  <br>
>
>정리하자면 call-by-value로 값을 복사받은 변수에 저장된 값을 변경한다고 해도 원래 변수에는 아무런 영향도 끼치지 못한다는 말이다.



<br>



## 2. Call By Reference



<br>



```cpp

void CallByReference(int&  reference) {

    reference = 20;

    cout << reference << endl;

    cout << &reference << endl;

}



int main() {

    int num = 10;

    CallByReference(num);

    cout << num << endl;

    cout << &num << endl;

    return 0;

}

```



<br>



+ 이렇게 하면 결과가 어떨까?



<br>



>20 <br>
>
>00000009910FFC44 <br>
>
>20 <br>
>
>00000009910FFC44



<br>



>이에서 알 수 있듯이, 참조자(&)를 이용한다는 것은 말 그대로 '참조'에 의한 전달을 하는 것으로, 위의 코드에서 살펴보면 reference 라는 변수는 num이 할당된 메모리를 공유하는 변수인 것이다 . <br>
>
> 위에서 언급했듯이 '별명'이라는 관점으로 생각하면, "CallByReference 함수 안에서는 'num'이라는 변수를 'reference'라고 부를거야." 정도로 생각하면 되겠다. <br>
>
> 그러니 reference와 num의 값, 주소를 출력했을 때 같은 값이 출력되는 것이 당연하다. <br>
>
> 따라서 원래 변수의 주소를 참조하는 변수의 값을 변경한다는 것은, 원래 변수의 값을 변경하겠다는 말과 똑같은 것이다.



<br>



## 3. Call By Address



<br>



```cpp

void CallByAddress(int*  address) {

    *address = 20;

    cout << *address << endl;

    cout << &address << endl;

}



int main() {

    int num = 10;

    CallByAddress(&num);

    cout << num << endl;

    cout << &num << endl;

    return 0;

}

```



<br>



+ 이번에는 포인터를 이용해보자. 어떤 결과가 출력될까?



<br>



>20 <br>
>
>0000007FE9D0F5D0 <br>
>
>20 <br>
>
>0000007FE9D0F5F4



<br>



>뭐지? 값은 변경되었는데, 두 변수의 주소는 다르다. <br>
>
>그 이유는, call-by-address 방식으로 생성된 포인터 변수 address는 본인만의 주소를 가지고 있으나, num의 주소를 저장하고 있기 때문에 역참조 연산자인 (* ) 를 이용하여 num의 주소에 접근하고 20 이라는 값을 저장하면 실제 num의 값 또한 바뀌게 되는 것 이다. <br>
>
> 정리하자면, call-by-address는 call-by-value 처럼 '변수의 주소'라는 '값'을 인자로 넘기는 방식이지만, call-by-reference와 같이 주소에 해당되는 값을 변경할 수도 있는 혼합된 방식이라고 생각하는 것이 좋겠다.



<br>



---

<br>

